File: Trackfield/app/src/main/java/com/example/trackfield/ui/custom/dialog/BinaryDialog.java
package com.example.trackfield.ui.custom.dialog;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import com.example.trackfield.R;
import com.example.trackfield.utils.annotations.Debug;

public class BinaryDialog extends BaseDialog {

    // bundle keys
    private static final String BUNDLE_PASS_VALUE = "passValue";

    // dialog tags
    private static final String TAG_DEFAULT = "binaryDialog";

    private DialogListener listener;

    // arguments
    @Nullable private String passValue = null;

    //

    public static BinaryDialog newInstance(@StringRes int titleRes, @StringRes int messageRes,
        @StringRes int posBtnTxtRes, String tag) {

        BinaryDialog instance = new BinaryDialog();
        Bundle bundle = putBundleBase(titleRes, messageRes, posBtnTxtRes, tag);
        instance.setArguments(bundle);

        return instance;
    }

    public static BinaryDialog newInstance(@StringRes int titleRes, @StringRes int messageRes,
        @StringRes int posBtnTxtRes, @Nullable String passValue, String tag) {

        BinaryDialog instance = new BinaryDialog();
        Bundle bundle = putBundleBase(titleRes, messageRes, posBtnTxtRes, tag);

        bundle.putString(BUNDLE_PASS_VALUE, passValue);

        instance.setArguments(bundle);
        return instance;
    }

    @Debug
    public static BinaryDialog generic(String tag) {
        return newInstance(R.string.dialog_title_continue, NO_RES, R.string.dialog_btn_continue, tag);
    }

    // extends DialogFragment

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        try {
            listener = (DialogListener) context;
        }
        catch (ClassCastException e) {
            throw new ClassCastException("Activity must implement DialogListener");
        }
    }

    // extends BaseDialog

    @Override
    protected void unpackBundle() {
        Bundle bundle = unpackBundleBase(TAG_DEFAULT);
        passValue = bundle.getString(BUNDLE_PASS_VALUE, null);
    }

    @Override
    protected AlertDialog buildDialog() {
        if (!message.equals("")) builder.setMessage(message);

        builder.setTitle(title)
            .setPositiveButton(posBtnTxtRes, (dialog, id) -> listener.onBinaryDialogPositiveClick(passValue, tag))
            .setNegativeButton(negBtnTxtRes, (dialog, id) -> getDialog().cancel());

        return builder.show();
    }

    // interface

    public interface DialogListener {

        void onBinaryDialogPositiveClick(String passValue, String tag);

    }

}

File: Trackfield/android/src/main/java/com/felwal/android/widget/dialog/BinaryDialog.kt
package com.felwal.android.widget.dialog

import android.content.Context
import android.os.Bundle
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.felwal.android.R

private const val ARG_PASS_VALUE = "passValue"

class BinaryDialog : BaseDialog() {

    private lateinit var listener: DialogListener

    // args
    private var passValue: String? = null

    // DialogFragment

    override fun onAttach(c: Context) {
        super.onAttach(c)

        listener = try {
            c as DialogListener
        }
        catch (e: ClassCastException) {
            throw ClassCastException("Activity must implement DialogListener")
        }
    }

    // BaseDialog

    override fun unpackBundle(bundle: Bundle?) {
        bundle?.apply {
            passValue = getString(ARG_PASS_VALUE, null)
        }
    }

    override fun buildDialog(): AlertDialog = builder.run {
        setTitle(title)
        if (message != "") setMessage(message)

        setPositiveButton(posBtnTxtRes) { _, _ ->
            listener.onBinaryDialogPositiveClick(passValue, dialogTag)
        }
        setCancelButton(negBtnTxtRes)

        show()
    }

    //

    interface DialogListener {
        fun onBinaryDialogPositiveClick(passValue: String?, tag: String)
    }

    //

    companion object {
        @JvmStatic
        fun newInstance(
            title: String,
            message: String = "",
            @StringRes posBtnTxtRes: Int = R.string.dialog_btn_ok,
            @StringRes negBtnTxtRes: Int = R.string.dialog_btn_cancel,
            tag: String,
            passValue: String? = null
        ): BinaryDialog = BinaryDialog().apply {
            arguments = putBaseBundle(title, message, posBtnTxtRes, negBtnTxtRes, tag).apply {
                putString(ARG_PASS_VALUE, passValue)
            }
        }
    }
}

fun binaryDialog(
    title: String,
    message: String = "",
    @StringRes posBtnTxtRes: Int = R.string.dialog_btn_ok,
    @StringRes negBtnTxtRes: Int = R.string.dialog_btn_cancel,
    tag: String,
    passValue: String? = null
): BinaryDialog = BinaryDialog.newInstance(title, message, posBtnTxtRes, negBtnTxtRes, tag, passValue)